quinn-proto/src/connection/mod.rs

```diff
@@ -1812,10 +1812,25 @@ impl Connection {
             false,
             false,
         );
+
         self.process_decrypted_packet(now, remote, Some(packet_number), packet)?;
         if let Some(data) = remaining {
             self.handle_coalesced(now, remote, ecn, data);
         }
+
+        if self.highest_space == SpaceId::Initial && self.state.is_handshake() {
+            // "The payload of an Initial packet includes a CRYPTO frame (or frames) containing a
+            // cryptographic handshake message, ACK frames, or both."
+            // https://www.rfc-editor.org/rfc/rfc9000.html#section-17.2.2
+            let space = &self.spaces[SpaceId::Initial];
+            if space.crypto_stream.bytes_read() == 0 {
```

Comment:
There might be Initial packets other than the first packet. Any reason not to check those too? We could place the logic in `process_early_payload`.

Reply:
Hmm, the spec specifically talks about the first packet:

> The first packet sent by a client always includes a CRYPTO frame that contains the start or all of the first cryptographic handshake message.

For other frames, it looks like PING is also allowed, for example.

Reply:
Ah, you're right.